Code/GraphMol/MolProcessing/MolProcessing.cpp

```diff
@@ -0,0 +1,156 @@
+//
+//  Copyright (C) 2024 Greg Landrum and other RDKit contributors
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include "Processing.h"
+
+namespace RDKit {
+namespace MolProcessing {
+
+namespace details {
+GeneralMolSupplier::SupplierOptions defaultSupplierOptions;
+inline std::unique_ptr<FileParsers::MolSupplier> getSupplier(
+    const std::string &fileName,
+    const GeneralMolSupplier::SupplierOptions &options) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+  static std::once_flag flag;
+  std::call_once(flag, []() { defaultSupplierOptions.numWriterThreads = 4; });
+#endif
+  return GeneralMolSupplier::getSupplier(fileName, options);
+}
+}  // namespace details
+
+namespace {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+inline std::mutex &fp_mutex_get() {
+  // create on demand
+  static std::mutex _mutex;
+  return _mutex;
+}
+
+inline void fp_mutex_create() {
+  std::mutex &mutex = fp_mutex_get();
+  std::lock_guard<std::mutex> test_lock(mutex);
+}
+
+inline std::mutex &get_fp_mutex() {
+  static std::once_flag flag;
+  std::call_once(flag, fp_mutex_create);
+  return fp_mutex_get();
+}
+
+template <typename T>
+void mtWorker(v2::FileParsers::MultithreadedMolSupplier *suppl,
+              std::function<T *(RWMol &)> func,
+              std::vector<std::unique_ptr<T>> &results) {
+  PRECONDITION(suppl, "no supplier");
```

Comment:
Perhaps adding a processing chunk size might be useful to protect against memory issues.  This would also remove the need for the mutex.

oid mtWorker(v2::FileParsers::MultithreadedMolSupplier *suppl,
               std::function<T *(RWMol &)> func,
               std::vector<std::unique_ptr<T>> &results,
               size_t procesCount, size_t startId) {
 std::vector<T*> accum(processCount);
 ...
    accum[recordId-startId] = func(mol);
 ...
 
although this may conflict with how suppl->next works in the threaded environment

Reply:
Here, memory issues are giving the processing system a billion molecules :)

Reply:
It's not going to misbehave any more than if you use the pattern with the molsupplier and try to store a billion fingerprints at once.

Reply:
If you use a normal mol supplier, you have more control though, don't you?  With this API, it appears to have to process the whole file before you get the results out.  I.e. you are iterating over molecules, not results on the output side.

Reply:
yes, this is a simple API that iterates over the whole supplier and gives you all the fingerprints at once.
It would be possible to use a callback that does something else (i.e. writes the FPs to a file), but that's not this particular function.